### PR TITLE
Production pull - extra validation on room creation endpoint & more permissions for developers.

### PR DESCRIPTION
**Pull request by bubby932**
* Grants users with the "Developer" tag every permission in every room.
* Adds more validation to the room creation endpoint.

### CODE REVIEW

File: routers/rooms.js [hunk 1]
```diff
@@ -171,7 +171,7 @@ router.get("/search", authenticateToken_optional, async (req, res) => {
     case "mine":
         return res.status(200).json(results.filter(room => room.creator_id === req.user.id));
     default:
-        return res.status(400).json({message: "invalid_mode"});
+        return res.status(400).json({code: "invalid_mode"});
```
Comment (bubby932) [inline routers/rooms.js:174]: This is a code, not a message.

File: routers/rooms.js [hunk 2]
```diff
@@ -453,11 +453,27 @@ router.post('/new', authenticateDeveloperToken, async (req, res) => {
     try {
         const { name } = req.body;
 
+        const coll = require('../index').mongoClient.db(process.env.MONGOOSE_DATABASE_NAME).collection("rooms");
```
Comment (bubby932) [inline routers/rooms.js:456]: Move declaration & definition higher.

File: routers/rooms.js [hunk 3]
```diff
@@ -453,11 +453,27 @@ router.post('/new', authenticateDeveloperToken, async (req, res) => {
...
+
+        
         if(typeof name != 'string') return res.status(400).json({
             code: "unspecified_parameter",
             message: "You did not specify the parameter 'name' in your request body."
         });
 
+        const predecessor = await coll.findOne(
```
Comment (bubby932) [inline routers/rooms.js:464]: Perform additional validation before continuing.

**Review by bubby932 — COMMENTED** [review]
Explained changes.

File: routers/rooms.js [hunk 4]
```diff
@@ -4,7 +4,7 @@ const Fuse = require('fuse.js');
 const express = require('express');
 const { getStorage } = require('firebase-admin/storage');
 const { v1 } = require('uuid');
-const { auditLog } = require('../helpers');
+const { auditLog, PullPlayerData } = require('../helpers');
```
Comment (bubby932) [inline routers/rooms.js:7]: Extra dependency

File: routers/rooms.js [hunk 5]
```diff
@@ -563,7 +577,7 @@ async function canViewRoom(req, res, next) {
     const rolePermissions = room.rolePermissions;
 
     const role = Object.keys(userPermissions).includes(req.user.id) ? userPermissions[req.user.id] : "everyone";
-    const canView = rolePermissions[role].viewAndJoin;
+    const canView = rolePermissions[role].viewAndJoin || req.user.developer;
```
Comment (bubby932) [inline routers/rooms.js:580]: Granted developers full view permissions.

File: routers/rooms.js [hunk 6]
```diff
@@ -583,7 +597,10 @@ async function hasPermission(user_id, room_id, permission) {
     var room = await client
         .db(process.env.MONGOOSE_DATABASE_NAME)
         .collection('rooms')
-        .findOne({_id: {$eq: room_id, $exists: true}});
+        .findOne({ _id: { $eq: room_id, $exists: true } });
```
Comment (bubby932) [inline routers/rooms.js:600]: Granted developers every permission.

**Review by bubby932 — COMMENTED** [review]
Updated previous explanation.

**Review by Bobrobot1 — APPROVED** [review]

**Comment by bubby932** [comment]
image
speed

**Pull request merged**